### ISSUE TITLE
QA: Repository Settings

### ISSUE THREAD

**Issue by Kreyren** [PhoenicisOrg/phoenicis#1756]
![image](https://user-images.githubusercontent.com/11302521/50951226-4d903800-14a4-11e9-903c-d090f659b84d.png)

### Expected behavior
1 -> 2 -> 3

### Actual behavior
3 -> 2 -> 1

### Steps to reproduce
Add 2 repositories in Settings/Repositories. 

### Reasoning
It's missleading

**Comment by Kreyren** [comment]
EDIT: Repositories should also overwrite each other if higger priority repo changes content. -> Allow branch patches

**Comment by Kreyren** [comment]
EDIT: Removed minor since it's complicating contributions.

**Comment by madoar** [comment]
I think you're describing multiple problems/suggestions in this issue. Maybe you should think about splitting them up into multiple GitHub issues to make it easier to track them and to comment to them :)

> Expected behavior
>> 1 -> 2 -> 3

So basically you suggest to switch the numbers (i.e top -> bottom increasing and not decreasing), right? When I implemented the repositories tab I thought about doing this. The reason why I didn't switch the order was, that the logic in my opinion is, that a higher priority for a normal user is also a higher number, right? (I know for developers this is normally not the case ;))
If we decide now that this is not the case, then I have no problem to change that.

> Repositories should also overwrite each other if higger priority repo changes content. -> Allow branch patches

Repositories already overwrite each other, just not completely.
When adding multiple repositories to Phoenicis, Phoenicis checks whether multiple repositories contain scripts for the same application. If this is the case Phoenicis selects the scripts from the repository with the higher priority.

**Comment by Kreyren** [comment]
@madoar I'm sugesting changing the numering from 1 at the top counting down. -> Seems more intuitive. -> Indicates what is loaded first, second, ..., last.

Assuming that numering above is true -> if 2nd is loaded it will overwrite 1st. -> if 3rd is loaded it won't overwrite difference between 1 and 2. -> allows patches for master repo using branches. 

difference between 1st and 2nd should also be kept. 

```
Repositories already overwrite each other, just not completely.
When adding multiple repositories to Phoenicis, Phoenicis checks whether multiple repositories contain scripts for the same application. If this is the case Phoenicis selects the scripts from the repository with the higher priority.
```

So 5 will overwrite 1,2,3,4 with difference in 5 ? or ?

**Comment by madoar** [comment]
Let me give an example. Let's assume you have two repositories added to Phoencisis:
1. The official repository
2. Your fork with an additional script, called `A` and a modified script, called `M`

Independent of the priorities of both repositories you'll see that `A` is shown in the application/script list of the application tab.

Additionally let us assume that your own fork has a higher priority than the official repository in Phoenicis (through dragging the repositories in the list).

Now you should see that the shown script for `M` in the application tab consists of the script located in your fork (i.e. the repository with the higher priority)

**Comment by Kreyren** [comment]
@madoar So if i have two repositories phoenicis is going to offer multiple installers for same wineapp?

**Comment by madoar** [comment]
What do you mean with `multiple installers for same wineapp`?
Phoenicis also supports multiple installation scripts for an application with only one repository added (see for example `Guild Wars 2` which has the two scripts `Online` and `Local`).

**Comment by Kreyren** [comment]
Now i'm even more confused O.O I understand from https://github.com/PhoenicisOrg/phoenicis/issues/1756#issuecomment-453059645 that it's going to load two options for two different sources of repository for end-user to chose?

**Comment by madoar** [comment]
No if there are two scripts with the same identifier from two different repositories, the script from the repository with the higher priority is taken.

**Comment by madoar** [comment]
There is an option in the settings of Phoenicis, which enables a tooltip for the `Install` buttons in the applications tab, that shows the source repository where the script is from, see `Settings -> User Interface Settings -> View the scripts’ source repository`.


To get a better feeling how Phoenicis chooses a script, when using multiple repositories, I recommend enabling this option and to play a bit around with different repository settings.

**Comment by Kreyren** [comment]
More info to `Settings -> User Interface Settings -> View the scripts’ source repository` ? I have that enabled, but i can't see the origin.

**Comment by Kreyren** [comment]
@madoar Recommends adding a description for the Repository settings, implementation is fine, but concept is hard to understand.

**Comment by Kreyren** [comment]
@madoar https://github.com/PhoenicisOrg/scripts/pull/797#issuecomment-454078055

Based on informations you've provide i would understand that it would load the installer from leagueoflegends branch? possible bug?

**Comment by Kreyren** [comment]
@madoar Also i stil ldon't see any change after `View the scripts's  source repository?`

![image](https://user-images.githubusercontent.com/11302521/51127753-90486c00-1826-11e9-9639-0ab1808c9a6f.png)

**Comment by plata** [comment]
The screenshot does not show your mouse so I can't tell what you want to show here. If you hover over "Install" (bottom right), it should work though.

**Comment by madoar** [comment]
Ok, I just took a look at the code, you need to hover your mouse over the name of the script, i.e. `Online` in your case.

**Comment by madoar** [comment]
@plata maybe we should think about changing the behavior of the tooltip?
I think it would be a good idea to show the tooltip if either the script name or the install button is hovered.

**Comment by Kreyren** [comment]
I would recommend trying some interactive GUI implementation with arrows and short description.

**Comment by plata** [comment]
Can you make a sketch? I cannot imagine what you have in mind currently.

**Comment by Kreyren** [comment]
> @madoar Also i still ldon't see any change after `View the scripts's source repository?`
> 
> ![image](https://user-images.githubusercontent.com/11302521/51127753-90486c00-1826-11e9-9639-0ab1808c9a6f.png)

Video: https://youtu.be/dPYRzItm-Uc

**Comment by madoar** [comment]
@Kreyren I can't reproduce your issue:
![show-script-source](https://user-images.githubusercontent.com/18488086/51173015-d2b88a00-18b4-11e9-950e-c629250f6d02.gif)

**Comment by Kreyren** [comment]
@madoar wierd, i tried standart theme and it works as expected, then i switched back on breeze dark theme and it also works..

**Comment by madoar** [comment]
About your suggestions:

> Replace numbers with arrows on both sides

I'm not sure if this helps understanding the function of the list better. I think it is enough to reverse the order of the numbers.

> Make the selection bigger by 25% ish

I don't understand what you mean?
Do you want to have all the text in the list 25% larger? Do you only want the selected row to be highlighted somehow?

> remove Repository

Agreed

> Shorten the selection of repositories by half

Why?

> make two big buttons add and remove and under them add refresh repository and restore defaults

This is basically what we already have. The only problem is basically the formatting (especially with `Add` an `Remove` being aligned the same as the repository list).
This is something that should likely be changed/fixed in css.
If you want to change this feel free to open a PR fixing this in the css :)

> Buttons should react to hover

What do you mean?
The hover behavior is defined in css. I guess it has been disabled in the dark breeze theme.
If you want to change this feel free to open a PR fixing this in the css :)

> Add description under or above it.

What description do you mean?

**Comment by madoar** [comment]
> Reversing the order of the numbers feels counter-productive. I believe that using arrows alike This overwrites this and that is way more effective.

@plata what do you think?

> Not text neccesary just the UI i believe it would make it more interactive.

You can do this by adjusting the scaling slider in the settings

> I'm not qualified to make the PR (i don't know where the source for it is and how to test it) + i can't fit it in my curent workload.

This is a cosmetic problem. Currently this has not a huge priority

> To explain how are the scripts pulled if more branches have same scripts?

This sounds like something, that should be shown in a tooltip.

**Comment by plata** [comment]
I understand the intention of the arrows but I'm not convinced that they will make things clearer (although the numbers are not intuitive either). What about a combobox where you can select the sorting ("priority (low to high)" and "priority (high to low)")?

**Comment by madoar** [comment]
This sounds like an application for a `TableView` component.
Maybe we should change our list to a table?

Just in case:
Drag and Drop in Tables: https://stackoverflow.com/questions/12353576/drag-drop-in-javafx-table?rq=1

**Comment by plata** [comment]
I'm thinking if I know any application with a similar functionality which would handle the sorting differently.

**Comment by madoar** [comment]
This issue has been partially solved by #1778.
The remaining tasks are tracked by the issues: #1819, #1818 and #1785.
